File: src/fantasypros.rs
use reqwest::header::ACCEPT;
use serde::Deserialize;
use std::error::Error;

#[derive(Deserialize)]
pub struct PlayerRank {
    pub player_id: i32,
    pub player_name: String,
    pub player_team_id: String,
    pub player_position_id: String,
    pub player_yahoo_id: String,
    pub rank_ecr: u32,
}

#[derive(Deserialize)]
pub struct PlayerRankings {
    pub players: Vec<PlayerRank>,
    pub last_updated: String,
}

pub struct FantasyPros {
    pub overall_rankings: PlayerRankings,
    pub superflex_rankings: PlayerRankings,
}

impl FantasyPros {
    pub fn new(api_key: String, year: i32) -> Result<FantasyPros, Box<dyn Error>> {
        // Superflex doesn't include D/ST or K
        let superflex_rankings = FantasyPros::get_ranks(&api_key, year, "OP")?;
        let overall_rankings = FantasyPros::get_ranks(&api_key, year, "ALL")?;
        Ok(FantasyPros {
            overall_rankings,
            superflex_rankings,
        })
    }
}

impl FantasyPros {
    fn get_ranks(
        api_key: &String,
        year: i32,
        position: &str,
    ) -> Result<PlayerRankings, Box<dyn Error>> {
        let url = format!("https://api.fantasypros.com/v2/json/nfl/{}/consensus-rankings?type=draft&scoring=PPR&position={}&week=0&experts=available", year, position);

        let client = reqwest::blocking::Client::new();

        let request = client
            .get(url)
            .header(ACCEPT, "application/json")
            .header("x-api-key", api_key);
        let response = request.send()?;

        // println!("FantasyPros:\n{}", response.text().unwrap());
        // Ok(PlayerRankings {
        //     players: vec![],
        //     last_updated: "".to_string(),
        // })

        let player_rankings = response.json::<PlayerRankings>()?;
        Ok(player_rankings)
    }
}

File: scraper/src/main/java/com/albersm/sleeperhelper/fantasypros/FantasyPros.java
package com.albersm.sleeperhelper.fantasypros;

import com.albersm.sleeperhelper.fantasypros.model.PlayerRankings;
import com.albersm.sleeperhelper.fantasypros.model.Rankings;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;

import static com.albersm.sleeperhelper.Scraper.OBJECT_MAPPER;

public class FantasyPros {

    private final Rankings rankings;
    private final HttpClient client = HttpClient.newHttpClient();

    public FantasyPros(String apiKey, int year) throws IOException, InterruptedException {
        var overallRankings = getRankings(apiKey, year, "ALL");
        var superFlexRankings = getRankings(apiKey, year, "OP");
        this.rankings = new Rankings(overallRankings, superFlexRankings);
    }

    public Rankings getRankings() {
        return rankings;
    }

    private PlayerRankings getRankings(String apikey, int year, String position) throws IOException, InterruptedException {
        var url = "https://api.fantasypros.com/v2/json/nfl/" + year
                + "/consensus-rankings?type=draft&scoring=PPR&position=" + position
                + "&week=0&experts=available";

        var uri  = URI.create(url);

        var request = HttpRequest.newBuilder()
                .uri(uri)
                .header("Accept", "application/json")
                .header("x-api-key", apikey)
                .build();

        var response = client.send(request, HttpResponse.BodyHandlers.ofString());

        var body = response.body();

        return OBJECT_MAPPER.readValue(body, PlayerRankings.class);
    }
}
